src/data/polynomial/iterated_deriv.lean

```diff
@@ -0,0 +1,234 @@
+end
+
+@[simp] lemma iterated_deriv_smul : iterated_deriv (r • p) n = r • iterated_deriv p n :=
+begin
+  induction n with n ih,
+  { simp only [iterated_deriv_zero_right] },
+  { simp only [iterated_deriv_succ, ih, derivative_smul] }
+end
+
+@[simp] lemma iterated_deriv_X_zero : iterated_deriv (X : polynomial R) 0 = X :=
+by simp only [iterated_deriv_zero_right]
+
+@[simp] lemma iterated_deriv_X_one : iterated_deriv (X : polynomial R) 1 = 1 :=
+by simp only [iterated_deriv, derivative_X, function.iterate_one]
+
+@[simp] lemma iterated_deriv_X (h : 1 < n) : iterated_deriv (X : polynomial R) n = 0 :=
+begin
+  induction n with n ih,
+  { exfalso, exact not_lt_zero 1 h},
+  { simp only [iterated_deriv_succ],
+    by_cases H : n = 1,
+    { rw H, simp only [iterated_deriv_X_one, derivative_one] },
+    { replace h : 1 < n := array.push_back_idx h (ne.symm H),
+      rw ih h, simp only [derivative_zero] } }
+end
+
+
+@[simp] lemma iterated_deriv_C_zero : iterated_deriv (C r) 0 = C r :=
+  by simp only [iterated_deriv_zero_right]
+
+@[simp] lemma iterated_deriv_C (h : 0 < n) : iterated_deriv (C r) n = 0 :=
+begin
+  induction n with n ih,
+  { exfalso, exact nat.lt_asymm h h },
+  { by_cases H : n = 0,
+    { rw [iterated_deriv_succ, H], simp only [iterated_deriv_C_zero, derivative_C]},
+    { replace h : 0 < n := nat.pos_of_ne_zero H,
+      rw [iterated_deriv_succ, ih h], simp only [derivative_zero] } }
+end
+
+@[simp] lemma iterated_deriv_one_zero : iterated_deriv (1 : polynomial R) 0 = 1 :=
+by simp only [iterated_deriv_zero_right]
+
+@[simp] lemma iterated_deriv_one : 0 < n → iterated_deriv (1 : polynomial R) n = 0 := λ h,
+begin
+  have eq1 : (1 : polynomial R) = C 1 := by simp only [ring_hom.map_one],
+  rw eq1, exact iterated_deriv_C _ _ h,
+end
+
+end semiring
+
+section ring
+variables [ring R] (p q : polynomial R) (n : ℕ)
+
+@[simp] lemma iterated_deriv_neg : iterated_deriv (-p) n = - iterated_deriv p n :=
+begin
+  induction n with n ih,
+  { simp only [iterated_deriv_zero_right] },
+  { simp only [iterated_deriv_succ, ih, derivative_neg] }
+end
+
+@[simp] lemma iterated_deriv_sub : iterated_deriv (p - q) n = iterated_deriv p n - iterated_deriv q n :=
+by rw [sub_eq_add_neg, iterated_deriv_add, iterated_deriv_neg, ←sub_eq_add_neg]
+
+
+end ring
+
+section comm_semiring
+variable [comm_semiring R]
+variables (f p q : polynomial R) (n k : ℕ)
+
+lemma coeff_iterated_deriv_as_prod_Ico :
+  ∀ m : ℕ, (iterated_deriv f k).coeff m = (∏ i in Ico m.succ (m + k.succ), i) * (f.coeff (m+k)) :=
+begin
+  induction k with k ih,
+  { simp only [add_zero, forall_const, one_mul, Ico.self_eq_empty, eq_self_iff_true,
+      iterated_deriv_zero_right, prod_empty] },
+  { intro m, rw [iterated_deriv_succ, coeff_derivative, ih (m+1), mul_assoc],
+    conv_lhs { congr, skip, rw mul_comm }, rw ←mul_assoc,
+    apply congr_arg2,
+    { have set_eq : (Ico m.succ (m + k.succ.succ)) = (Ico (m + 1).succ (m + 1 + k.succ)) ∪ {m+1},
```

Comment:
imo this should be a lemma in the library, but nat intervals barely has any API at the moment. I'm happy for this to be how it is right now.